conjure-java-jersey-server/src/main/java/com/palantir/conjure/java/server/jersey/BearerTokenParamConverterProvider.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * (c) Copyright 2017 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.conjure.java.server.jersey;
+
+import com.palantir.logsafe.Preconditions;
+import com.palantir.tokens.auth.BearerToken;
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Type;
+import javax.ws.rs.ext.ParamConverter;
+import javax.ws.rs.ext.ParamConverterProvider;
+import javax.ws.rs.ext.Provider;
+import org.glassfish.jersey.internal.inject.Custom;
+
+@Custom
+@Provider
+public final class BearerTokenParamConverterProvider implements ParamConverterProvider {
```

Comment:
I'm not confident this is correct, while cookie auth uses the BearerToken type, we can define additional parameters of type BearerToken that aren't the conjure-defined auth token.

Reply:
I mention this in the PR description:
> These param converters will apply to all `AuthHeader` and `BearerToken` parameters. However it is very rare for these types to be used outside of the endpoint auth. And most of those cases are to perform some sort of delegate authentication, so a 401 is not inappropriate.

I don't think I've ever seen a `BearerToken` used as an argument outside of cookie auth. And the only times I've seen a `AuthHeader` used as argument outside of header auth is for delegating authorization, in which case you could argue that a 401 is not inappropriate.

I don't think there is another way to achieve the desired behavior. If you have other suggestions, I'm happy to hear them. But I think the benefit of responding with 401 for invalid auth credentials outweighs the minor downside of this applying to other parameters of the same type.

Reply:
I've updated this PR to only apply to parameters that could be Conjure auth parameters. This means:
- `AuthHeader` parameter must be annotated with `@HeaderParam` and have a value of `Authorization`
- `BearerToken` parameter must be annotated with `@CookieParam` and have any value

Reply:
That sounds reasonable, I don't think we need the additional specificity for AuthHeader because conjure only uses that type for auth components, bearer tokens use the `BearerToken` type.

Reply:
I would prefer to not respond with 401 for parameters we know aren't the conjure auth parameter.

We actually do have a number of endpoints with a second `@HeaderParam` `AuthHeader` that represents a delegating user. It makes more sense to return 400 for these parameters since they are effectively arguments that just happen to be passed as headers.